src/Hyperion/Benchmark.hs

```diff
@@ -55,7 +60,37 @@ instance Show Benchmark where
   showsPrec x (WithSampling _opt bk) = showsPrec x bk
 
 bench :: String -> Batch () -> Benchmark
-bench name batch = Bench (Text.pack name) batch
+bench name batch = Bench Text.pack (const []) id (const batch) name
+
+benchWith
+  :: (a -> Text)
+  -> (a -> [JSON.Pair])
```

Comment:
I understand that the user might want to attach arbitrary metadata. But this type is breaking an abstraction barrier. We shouldn't assume in the benchmarks DSL that the output is always JSON. Indeed it might not be. The DSL should be independent of the output format for the results of the benchmark.

So there needs to be a very good reason to justify breaking this abstraction barrier. Could you give me specific examples of actual metadata you want to attach? Metadata that is not already covered by the existing "series" functionality? I suspect that rather than a general attach-arbitrary-user-metadata mechanism, we can get away with supporting a fixed set of metadata, supporting pretty much all known use cases. In this way we'd avoid having to allow the user to inject arbitrary JSON key-value pairs.